src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+
+#include <base_alloc.h>
+#include <base_alloc_linear.h>
+#include <base_alloc_os_mem.h>
+
+// pool size of the base allocator,
+// it will be reallocated if we run out of memory.
+#define BA_POOL_SIZE (ba_os_get_page_size())
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct ba_list_item_t ba_list_item_t;
+
+struct ba_list_item_t {
+    ba_list_item_t *next;
+    void *addr;
+};
+
+struct umf_ba_pool_t {
+    size_t aligned_size;
+    ba_list_item_t *free_list;
+    ba_list_item_t *alloc_list;
+#ifdef DEBUG
+    long tid;
+#endif /* DEBUG */
+};
+
+static int ba_fill_pool(umf_ba_pool_t *pool) {
+    assert(pool->free_list == NULL);
+
+    char *pool_memory = ba_os_alloc(BA_POOL_SIZE);
+    if (!pool_memory) {
+        return -1;
+    }
+
+    size_t pool_size_left = BA_POOL_SIZE;
+    ba_list_item_t *last_list_item = NULL;
+
+    while (pool_size_left >= pool->aligned_size) {
+        ba_list_item_t *new_list_item =
+            (ba_list_item_t *)linear_alloc(sizeof(ba_list_item_t));
+        if (!new_list_item) {
+            return -1;
+        }
+
+        new_list_item->next = last_list_item;
+        new_list_item->addr = pool_memory;
+
+        last_list_item = new_list_item;
+        pool_size_left -= pool->aligned_size;
+        pool_memory += pool->aligned_size;
+    }
+
+    pool->free_list = last_list_item;
+
+    return 0;
+}
+
+umf_ba_pool_t *umf_ba_create(size_t size) {
+    size_t aligned_size;
+    // align size to MEMORY_ALIGNMENT bytes
+    size_t rest = size & (MEMORY_ALIGNMENT - 1);
+    if (rest) {
+        aligned_size = size - rest + MEMORY_ALIGNMENT;
+    } else {
+        aligned_size = size;
+    }
+
+    umf_ba_pool_t *pool = (umf_ba_pool_t *)linear_alloc(sizeof(umf_ba_pool_t));
+    if (!pool) {
+        return NULL;
+    }
+
+    pool->aligned_size = aligned_size;
+    pool->alloc_list = NULL;
+    pool->free_list = NULL;
+#ifdef DEBUG
+    pool->tid = ba_os_get_tid();
+#endif /* DEBUG */
+
+    if (ba_fill_pool(pool)) {
+        return NULL;
+    }
+
+    return pool;
+}
+
+void *umf_ba_malloc(umf_ba_pool_t *pool) {
+    // a BA pool can be used only in one thread
+#ifdef DEBUG
+    assert(pool->tid == ba_os_get_tid());
+#endif /* DEBUG */
+
+    if (!pool->free_list) {
+        if (ba_fill_pool(pool)) {
+            return NULL;
+        }
+    }
+
+    ba_list_item_t *item = pool->free_list;
+    pool->free_list = pool->free_list->next;
+    item->next = pool->alloc_list;
+    pool->alloc_list = item;
+
+    assert(item->addr);
+    void *ptr = item->addr;
+    item->addr = NULL;
+
+    return ptr;
+}
+
+void umf_ba_free(umf_ba_pool_t *pool, void *ptr) {
```

Comment:
I don't think we can forbid user from calling free from a different thread (than the one used for allocating `ptr`). We will need to support this case for critnib and memory tracking.

Reply:
It seems so. I've just dreamt I could make it lock-free ;-) ...

Reply:
Done

Reply:
I think we could only have a mutex around the `ba_os_alloc` call and all `next` and `free_list` pointers modifications could be done using compare exchange. Free could also be lock-free then.

Of course we can do it in another PR

Reply:
I think we could just remove this assert in `ba_divide_memory_into_chunks` and once all the memory is divided just do compare exchange on pool->metadata.free_list. If it's empty, it will be equivalent to current logic, if it's not, then this would be similar to how we would insert to the list in `umf_ba_free`.

So, there would be 3 CAS operations:
- in umf_ba_free() when inserting an item to the list
- in umf_ba_alloc() when taking an item from the list
- in ba_divide_memory_into_chunks() when inserting a bunch items to the list

Reply:
@igchor Is that what you meant?
https://github.com/ldorau/unified-memory-framework/commit/ef34eef7058b4a150a8bcc7b49d1a6d9bf6294c4

Reply:
Yes, exactly. 

I would only propose implementing an os-agnostic version of `__sync_bool_compare_and_swap` in utils_concurreny.h and using that. Also when you're reading the old `free_list` value you should use util_atomic_load_acquire.

Reply:
And there's also a question whether we should get rid of the mutex around `ba_os_alloc` in `umf_ba_alloc`.  Without the mutex, if `pool->metadata.free_list == NULL` and 2 (or more) threads happen to call `umf_ba_alloc` at the same time, we can have multiple `ba_os_alloc` calls. Everything should be consistent but we'll possibly allocate more than needed.

I think it would be better to leave the mutex there. (the compare exchange operations are still needed to synchronize with free).

Reply:
One more thing: the `if (pool->metadata.free_list == NULL)` check inside `umf_ba_alloc` should probably be moved to the `do {} while` loop because some other thread might allocate the last free chunk from the free list while our thread is trying to do CAS.

Reply:
OK, so let's do that in a separate PR.

Reply:
New version: https://github.com/ldorau/unified-memory-framework/commit/f88053c040377613fcef64a64dfdeea0b2216740 - will be submitted in a separate PR